[E2E][block-widgets-editor] Add regression e2e test for missing visibility option for legacy widgets

fullofcaffeine opened the pull request:
#### Changes proposed in this Pull Request

Verify that the visibility button is present and works for legacy widgets inserted in the new block widgets editor. Only one of the block widgets is tested against and it's assumed that if it's available for it, it'll available for all others. 

This is also the first time I'm writing a Playwright test and working with its API, so bear with me :). Also, the name of the test is provisory and we might make it more specific or change it altogether, I'm open to suggestions.

This test couldn't be moved upstream (Jetpack or core) because the problem is specific to WPCOM's idiosyncrasies. More details in [this PR](https://github.com/Automattic/jetpack/pull/20644).

#### Testing instructions

* Tests should pass.

#### TODO
- [x] Make sure this also works well or handles the mobile viewport

Related to https://github.com/Automattic/jetpack/pull/20644.

fullofcaffeine committed:
Add test that verifies that the visibility option for legacy widgets is present

fullofcaffeine commented on packages/calypso-e2e/src/lib/components/sidebar-component.ts:
```diff
@@ -96,6 +107,8 @@ export class SidebarComponent {
 			await this._click( selector );
 		}
 
+		if ( skipWaitForContent ) return;
```
Had to add this as the `waitForSelector` was timing-out in the customizer->widgets page.

worldomonation replied:
Interesting. I think that perhaps it may make sense to consider removing Line 113 to simplify the method signature at some point. I think we can merge this as-is for now, since removing Line 113 will be a change that will impact all other specs and we don't need to do that now.

fullofcaffeine committed:
Slight wording improvement

fullofcaffeine commented on test/e2e/specs/specs-playwright/wp-widgets__spec.ts:
```diff
@@ -0,0 +1,50 @@
+import { DataHelper, LoginFlow, SidebarComponent, setupHooks } from '@automattic/calypso-e2e';
+import { Page } from 'playwright';
+
+const gutenbergUser =
```
I think we might not need that at all in this test, as this is not really GB specific - and what caused the bug wasn't really related to GB, though, so I might as well just remove it.

worldomonation replied:
I don't understand fully what and how the Gutenberg tests in TC are set up, so I might miss my mark here. Feel free to ignore if I don't make any sense 😆 

I think it would be best to move away from the practice that's seen often in the Selenium suite, to use a different user depending on the environment. I think it might make more sense to [parametrize the test](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/specs-playwright/wp-media__upload-spec.js#L24-L27) and test it under both `gutenbergSimpeSiteEdgeUser` and `gutenbergSimpleSiteUser`.

This way we have a proper baseline that isn't dependent on an environment variable being set.

fullofcaffeine replied:
I like that pattern! I'll keep that in mind for any eventual playwright GB tests I might write in the future. For now, this ended up not being needed at all, as this test is not really GB-specific :)

worldomonation commented on test/e2e/specs/specs-playwright/wp-widgets__spec.ts:
```diff
@@ -0,0 +1,50 @@
+		await loginFlow.logIn();
+	} );
+
+	// @todo: Refactor/Abstract these steps into a WidgetsEditor component
```
I can help with this part once you merge this PR :)

fullofcaffeine replied:
🎉

worldomonation commented on test/e2e/tsconfig.json:
```diff
@@ -2,7 +2,7 @@
 	"compilerOptions": {
 		"noEmit": true, // just type checking, no output. The output is handled by babel.
-        "types": [ "jest" ] // no mocha - we are only using TypeScript for the new Playwright scripts
+        "types": [ "jest", "node" ] // no mocha - we are only using TypeScript for the new Playwright scripts
```
I don't know enough about TypeScript configuration side of things - what does this do?

fullofcaffeine replied:
Type definitions for the node: https://www.npmjs.com/package/@types/node. This was needed for the `process` global. I will remove that line, so I can remove this too.

worldomonation replied:
Gotcha. Thanks for clarifying!

worldomonation commented on test/e2e/specs/specs-playwright/wp-widgets__spec.ts:
```diff
@@ -0,0 +1,50 @@
+		} );
+	} );
+
+	it( 'Dismiss the Welcome Gudie Notice', async function () {
```
non-blocker: Since this step may be optional, it may be a good idea to name this step something like 

`Dismiss the Welcome Guide if present`.

Also, typo in the word `Guide`.

fullofcaffeine replied:
Yeah, I had it with the `if present` suffix, though I found out it is always shown.

fullofcaffeine replied:
Not sure if it's a bug though, it appears it should be a one-time thing, so I'd still keep the condition that checks for the button meanwhile. I'll ask around to find out what's the correct behavior here.

worldomonation replied:
Sounds good! Not a blocker by any means.

worldomonation commented on test/e2e/specs/specs-playwright/wp-widgets__spec.ts:
```diff
@@ -0,0 +1,50 @@
+
+	it( 'Dismiss the Welcome Gudie Notice', async function () {
+		await page.waitForLoadState( 'networkidle' );
+		const btnSelector = 'button:text("Got it")';
```
I'd prefer that this constant either be named
`button`
or
`elementHandle`.

Stylistically, it is preferable to not contract names like in this case.

fullofcaffeine replied:
Sounds good 👍🏻

worldomonation commented:
Generally looks good! A few small nits and non-blockers. I appreciate that your step names avoid the use of modal verbs and are quite similar to what I would have used if I was writing the test.

fullofcaffeine committed:
Remove GB edge user/site switching logic, use the `defaultUser` instead

Also: fix typo.

fullofcaffeine committed:
Use the `gutenbergSimpleSiteUser` account

The `defaultUser` account uses the `e2eflowtest3` site, which is used for theme testing as well. This is problematic because themes change the behavior of the customizer and can make this test fail. More specifically, themes with more than 1 widget area have a slight different UI when you navigate to the customizer with the widgets pane in focus. Instead of opening the block editor by default, it shows each area as an accordion button, and that breaks the logic.

Someday, it might make sense to handle that scenario (of > 1 widgets area) in this test, but for the regression test goals we have now, it wouldn't be needed. We're instead going to use a site that we know has a theme with only 1 widget area, and that is not used to test themes.

fullofcaffeine committed:
Remove the `@types/node` package

fullofcaffeine commented on test/e2e/package.json:
```diff
@@ -27,10 +27,11 @@
 		"asana-phrase": "^0.0.8",
 		"cached-path-relative": ">=1.0.2",
-		"chromedriver": "^89.0.0",
+		"chromedriver": "^92.0.0",
```
OOpsie.

fullofcaffeine committed:
Improve const name, revert back to chromedriver 89

fullofcaffeine committed:
Revert yarn.lock changes

WunderBart committed:
Do not attempt to retry if we've navigated outside wp.com

fullofcaffeine commented on packages/calypso-e2e/src/lib/components/sidebar-component.ts:
```diff
@@ -63,6 +63,15 @@ export class SidebarComponent {
+		 * is no longer present.
+		 */
+		const currentURL = await this.page.url();
+		if ( ! currentURL.startsWith( 'https://wordpress.com' ) ) {
```
I think this is a good trade-off, but we should be aware that it'll eventually prevent us from running CI tests against a different URL, not a huge deal now as we haven't really done that yet, but there is a parameter for that in the `WPCom Tests->Gutenberg tests` `buildType`, called `Test URL`. cc @scinos

fullofcaffeine commented on packages/calypso-e2e/src/lib/components/sidebar-component.ts:
```diff
@@ -48,25 +48,43 @@ export class SidebarComponent {
+		 * is no longer present.
+		 */
+		const currentURL = await this.page.url();
+		if ( ! currentURL.startsWith( 'https://wordpress.com' ) ) {
```
I think this is a good trade-off, but we should be aware that it'll eventually prevent us from running CI tests against a different URL, not a huge deal now as we haven't really done that yet, but there is a parameter for that in the WPCom Tests->Gutenberg tests buildType, called Test URL. This could be a problem if we decide to migrate this param over to the Playwright buildTypes. cc @scinos

WunderBart replied:
Yeah, that value would need to be env-dependant. I'm wondering if we still need the retry logic though (after the latest sidebar tweaks), so maybe we can revisit that and in case it's not, this check will be gone as well.
